### PR TITLE
Make the coefficient material property for array time derivative kernel optional

### LINKED ISSUE #22830: Allow ArrayTimeDerivative to not have a coefficient
## Reason
Currently `ArrayTimeDerivative` requires a parameter for specifying a material property that to be multiplied in the kernel. But there are cases that the time derivative term does not have a coefficient. Making this parameter optional is thus desired.

## Design
Make the parameter unrequired and add one more switch in the residual evaluation for the kernel.

## Impact
A new feature that should not have impact.

### PR DESCRIPTION
**Pull request by YaqiWang**
Close #22830.

This small enhancement is needed for MSR transient calculations.

### CODE REVIEW

File: framework/src/kernels/ArrayReaction.C [hunk 1]
```diff
@@ -15,9 +15,9 @@ InputParameters
 ArrayReaction::validParams()
 {
   InputParameters params = ArrayKernel::validParams();
-  params.addParam<MaterialPropertyName>("reaction_coefficient",
-                                        "The name of the reactivity, "
-                                        "can be scalar, vector, or matrix.");
+  params.addRequiredParam<MaterialPropertyName>("reaction_coefficient",
```
Comment (YaqiWang) [inline framework/src/kernels/ArrayReaction.C:18]: This parameter is supposed to be required.

**Comment by YaqiWang** [comment]
This should be ready for review.

File: framework/src/kernels/ArrayTimeDerivative.C [hunk 2]
```diff
@@ -17,25 +17,28 @@ ArrayTimeDerivative::validParams()
   InputParameters params = ArrayTimeKernel::validParams();
   params.addClassDescription("Array time derivative operator with the weak form of $(\\psi_i, "
                              "\\frac{\\partial u_h}{\\partial t})$.");
-  params.addRequiredParam<MaterialPropertyName>("time_derivative_coefficient",
-                                                "The name of the time derivative coefficient. "
-                                                "Can be scalar, vector, or matrix");
+  params.addParam<MaterialPropertyName>("time_derivative_coefficient",
+                                        "The name of the time derivative coefficient. "
+                                        "Can be scalar, vector, or matrix");
```
Comment (GiudGiud) [inline framework/src/kernels/ArrayTimeDerivative.C:22]:
```suggestion
                                        "Can be scalar, vector, or matrix material property");
```

File: framework/src/kernels/ArrayReaction.C [hunk 3]
```diff
@@ -15,9 +15,9 @@ InputParameters
...
+                                                "The name of the reactivity, "
+                                                "can be scalar, vector, or matrix.");
```
Comment (GiudGiud) [inline framework/src/kernels/ArrayReaction.C:20]:
```suggestion
                                                "can be scalar, vector, or matrix material property.");
```

**Comment by GiudGiud** [comment]
I'm ok with this. If this is for precursors advection, we have great models with finite volume available please let us know.

**Comment by YaqiWang** [comment]
>  If this is for precursors advection, we have great models with finite volume available please let us know.

It is. I suggested that to Mustafa too. But his model is very simple, so using a Griffin built-in action is good enough to him.

**Review by GiudGiud — APPROVED** [review]

**Pull request merged**